server/ServerFSM.cpp

```diff
@@ -241,50 +241,79 @@ ServerApp& ServerFSM::Server()
         m_server.Networking().UpdateCookie(player_connection->Cookie());
 
-        int id = player_connection->PlayerID();
+        id = player_connection->PlayerID();
         DebugLogger(FSM) << "ServerFSM::HandleNonLobbyDisconnection : Lost connection to player #" << id
                          << ", named \"" << player_connection->PlayerName() << "\".";
 
-        // Did an active player (AI or Human) disconnect?  If so, game is over
-        if (player_connection->GetClientType() == Networking::CLIENT_TYPE_HUMAN_OBSERVER ||
-            player_connection->GetClientType() == Networking::CLIENT_TYPE_HUMAN_MODERATOR)
-        {
-            // can continue.  Select new host if necessary.
-            if (m_server.m_networking.PlayerIsHost(id))
-                m_server.SelectNewHost();
-
-        } else if (player_connection->GetClientType() == Networking::CLIENT_TYPE_HUMAN_PLAYER ||
-                   player_connection->GetClientType() == Networking::CLIENT_TYPE_AI_PLAYER)
-        {
+        if (player_connection->GetClientType() == Networking::CLIENT_TYPE_AI_PLAYER) {
+            // AI could safely disconnect only if empire was eliminated
             const Empire* empire = GetEmpire(m_server.PlayerEmpireID(id));
-            // eliminated and non-empire players can leave safely
             if (empire && !empire->Eliminated()) {
                 must_quit = true;
+                // AI abnormally disconnected during a regular game
+                ErrorLogger(FSM) << "AI Player #" << id << ", named \""
+                                 << player_connection->PlayerName() << "\"quit before empire was eliminated.";
+            }
+        } else if (player_connection->GetClientType() == Networking::CLIENT_TYPE_HUMAN_PLAYER) {
+            const Empire* empire = GetEmpire(m_server.PlayerEmpireID(id));
+            // eliminated and non-empire players can leave safely
+            if (empire && !empire->Eliminated()) {
                 // player abnormally disconnected during a regular game
-                ErrorLogger(FSM) << "Player #" << id << ", named \""
+                WarnLogger(FSM) << "Player #" << id << ", named \""
                                  << player_connection->PlayerName() << "\"quit before empire was eliminated.";
-            } else if (player_connection->GetClientType() == Networking::CLIENT_TYPE_HUMAN_PLAYER && !empire) {
-                // should be in ingame lobby
-                // update ingame lobby
-                UpdateIngameLobby();
+                // detach player from empire
+                m_server.DropPlayerEmpireLink(id);
             }
         }
     } else {
         DebugLogger(FSM) << "Client quit before id was assigned.";
     }
 
-    // independently of everything else, if there are no humans left, it's time to terminate
-    if (m_server.m_networking.empty()
-        || m_server.m_ai_client_processes.size() == m_server.m_networking.NumEstablishedPlayers())
+    int connected_count = 0;
```

Comment:
add a comment here explaining what is being tallied

Reply:
also, might this be better described as a count of the empires with connected players and of empires without a connected player?

Reply:
It's a count of active (non-eliminated) empires, which currently have a connected human players. Eliminated or AI players don't count.

Reply:
I'd give these variables clearer names, then

Reply:
Changed names and added comments.